subprojects/dependency-management/src/main/java/org/gradle/internal/resolve/caching/ComponentMetadataRuleExecutor.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.resolve.caching;
+
+import org.gradle.api.Transformer;
+import org.gradle.api.artifacts.ComponentMetadataContext;
+import org.gradle.api.artifacts.ModuleVersionIdentifier;
+import org.gradle.api.artifacts.ResolvedModuleVersion;
+import org.gradle.api.internal.artifacts.configurations.dynamicversion.CachePolicy;
+import org.gradle.api.internal.changedetection.state.InMemoryCacheDecoratorFactory;
+import org.gradle.api.internal.changedetection.state.ValueSnapshotter;
+import org.gradle.cache.CacheRepository;
+import org.gradle.internal.component.external.model.ModuleComponentResolveMetadata;
+import org.gradle.internal.serialize.Serializer;
+import org.gradle.util.BuildCommencedTimeProvider;
+
+import java.io.Serializable;
+
+public class ComponentMetadataRuleExecutor extends CrossBuildCachingRuleExecutor<ModuleComponentResolveMetadata, ComponentMetadataContext, ModuleComponentResolveMetadata> {
+
+    private final static Transformer<Serializable, ModuleComponentResolveMetadata> KEY_TO_SNAPSHOTTABLE = new Transformer<Serializable, ModuleComponentResolveMetadata>() {
+        @Override
+        public Serializable transform(ModuleComponentResolveMetadata moduleMetadata) {
+            return moduleMetadata.getContentHash().asBigInteger();
```

Comment:
:o: Using a `BigInteger` means that we will not used an optimized serializer (see ValueSnapshotter#processValue). Not necessarily a big deal in the beginning.

Reply:
There are actually 2 things:

1. the snapshotting, that could use have a specialized `BigIntegerValueSnapshot` or something like that
2. the key is serialized to disk. And because it's a `BigInteger`, and that there's no specialized snapshot, there's no special case in `SnapshotSerializer` and it will rely on Java serialization for all keys. Which is probably not that good.

Reply:
On the other hand if we move to the new `Work<I>` API, this is not going to be required anymore.

Reply:
Not so sure, this still is a more efficient key than the full `ModuleComponentResolveMetadata`